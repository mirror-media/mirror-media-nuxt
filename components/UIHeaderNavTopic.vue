<template>
  <section class="header-nav-topic">
    <div class="container">
      <div class="topic-container">
        <nuxt-link
          v-for="topic in topics"
          :key="topic.id"
          class="topic topic--normal"
          :to="`/topic/${topic.id}`"
          @click.native="emitGA(`topic ${topic.name}`)"
        >
          <h2>{{ topic.name }}</h2>
        </nuxt-link>
      </div>

      <nuxt-link
        to="/section/topic"
        class="topic"
        @click.native="emitGA('topic 更多')"
      >
        <h2>更多</h2>
      </nuxt-link>

      <div class="sub-brand-container">
        <a
          href="https://voice.mirrorfiction.com/"
          target="_blank"
          @click="emitGA('section mirrorvoice')"
        >
          <img src="~/assets/mirrorvoice.png" alt="鏡好聽" />
        </a>
        <a
          href="https://www.mirrorfiction.com/"
          target="_blank"
          @click="emitGA('section mirrorfiction')"
        >
          <img src="~/assets/mirrorfiction.png" alt="鏡文學" />
        </a>
        <a
          href="https://www.readr.tw/"
          target="_blank"
          @click="emitGA('section readr')"
        >
          <img src="~/assets/readrlogo-gary.png" alt="READr 讀+" />
        </a>
      </div>
    </div>
  </section>
</template>

<script>
export default {
  name: 'UIHeaderNavTopic',
  props: {
    topics: {
      type: Array,
      required: true,
      default: () => [],
    },
  },
  methods: {
    emitGA(eventLabel) {
      this.$emit('sendGA', {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel,
      })
    },
  },
}
</script>

<style lang="scss" scoped>
.header-nav-topic {
  display: none;
  @include media-breakpoint-up(xl) {
    display: block;
    background-color: #fff;
    color: rgba(#000, 0.5);
    text-align: center;
  }
  .container {
    max-width: 1024px;
    height: 45px;
    overflow: hidden;
    display: flex;
    margin-left: auto;
    margin-right: auto;
    user-select: none;
  }
  .topic-container {
    display: flex;
    flex-wrap: wrap;
    max-width: 582px;
  }
  .topic {
    min-width: 90px;
    height: 100%;
    padding-left: 8px;
    padding-right: 8px;
    line-height: 42px;
    cursor: pointer;
    &--normal {
      border-bottom: 3px solid #000;
      flex: 0 0 auto;
      &:hover {
        color: #fff;
        background-color: #000;
      }
    }
  }
  .sub-brand-container {
    display: flex;
    align-items: center;
    flex: 0 0 auto;
    & a {
      padding-left: 10px;
      padding-right: 10px;
      flex: 0 0 auto;
      cursor: pointer;
    }
    & img {
      height: 30px;
      width: auto;
    }
  }
}
</style>
