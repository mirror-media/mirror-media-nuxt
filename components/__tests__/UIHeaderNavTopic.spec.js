import { shallowMount } from '@vue/test-utils'
import UIHeaderNavTopic from '../UIHeaderNavTopic.vue'

const topic = {
  name: '人造地獄',
  id: '5d8c26c23c4cad5f87630d37',
}

describe('topic nav', () => {
  const wrapper = shallowMount(UIHeaderNavTopic, {
    propsData: {
      topics: [topic],
    },
    stubs: ['nuxt-link'],
  })

  test('render the proper topic link', () => {
    const linkNormal = wrapper.find(`[to="/topic/${topic.id}"]`)
    expect(linkNormal.find('h2').text()).toBe(topic.name)

    const linkMore = wrapper.find('[to="/section/topic"]')
    expect(linkMore.find('h2').text()).toBe('更多')
  })
})

describe('sub-brand nav', () => {
  const wrapper = shallowMount(UIHeaderNavTopic, {
    propsData: {
      topics: [],
    },
    stubs: ['nuxt-link'],
  })

  test('render the proper sub-brand link', () => {
    const voiceLink = wrapper.find('[href="https://voice.mirrorfiction.com/"]')
    expect(voiceLink.attributes().target).toBe('_blank')
    expect(voiceLink.find('img').attributes().alt).toBe('鏡好聽')

    const fictionLink = wrapper.find('[href="https://www.mirrorfiction.com/"]')
    expect(fictionLink.attributes().target).toBe('_blank')
    expect(fictionLink.find('img').attributes().alt).toBe('鏡文學')

    const readrLink = wrapper.find('[href="https://www.readr.tw/"]')
    expect(readrLink.attributes().target).toBe('_blank')
    expect(readrLink.find('img').attributes().alt).toBe('READr 讀+')
  })
})

describe('emitGA', () => {
  const wrapper = shallowMount(UIHeaderNavTopic, {
    propsData: {
      topics: [topic],
    },
    stubs: ['nuxt-link'],
  })

  beforeEach(function resetEmitted() {
    wrapper._emitted.sendGA = []
  })

  test('with a proper argument when click topic link', () => {
    const linkNormal = wrapper.find(`[to="/topic/${topic.id}"]`)
    linkNormal.trigger('click')
    expect(wrapper.emitted().sendGA[0]).toEqual([
      {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel: `topic ${topic.name}`,
      },
    ])

    const linkMore = wrapper.find(`[to="/section/topic"]`)
    linkMore.trigger('click')
    expect(wrapper.emitted().sendGA[1]).toEqual([
      {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel: 'topic 更多',
      },
    ])
  })

  test('with a proper argument when click sub-brand link', () => {
    const voiceLink = wrapper.find('[href="https://voice.mirrorfiction.com/"]')
    voiceLink.trigger('click')
    expect(wrapper.emitted().sendGA[0]).toEqual([
      {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel: 'section mirrorvoice',
      },
    ])

    const fictionLink = wrapper.find('[href="https://www.mirrorfiction.com/"]')
    fictionLink.trigger('click')
    expect(wrapper.emitted().sendGA[1]).toEqual([
      {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel: 'section mirrorfiction',
      },
    ])

    const readrLink = wrapper.find('[href="https://www.readr.tw/"]')
    readrLink.trigger('click')
    expect(wrapper.emitted().sendGA[2]).toEqual([
      {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel: 'section readr',
      },
    ])
  })
})
