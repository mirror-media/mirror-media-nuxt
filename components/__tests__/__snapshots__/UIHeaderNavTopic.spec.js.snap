// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markup render correctly 1`] = `
<section
  class="header-nav-topic"
>
  <div
    class="container"
  >
    <div
      class="topic-container"
    >
      <nuxt-link-stub
        class="topic topic--normal"
        to="/topic/5d8c26c23c4cad5f87630d37"
      >
        <h2>
          人造地獄
        </h2>
      </nuxt-link-stub>
    </div>
     
    <nuxt-link-stub
      class="topic"
      to="/section/topic"
    >
      <h2>
        更多
      </h2>
    </nuxt-link-stub>
     
    <div
      class="sub-brand-container"
    >
      <a
        href="https://voice.mirrorfiction.com/"
        target="_blank"
      >
        <img
          alt="鏡好聽"
          src="~/assets/mirrorvoice.png"
        />
      </a>
       
      <a
        href="https://www.mirrorfiction.com/"
        target="_blank"
      >
        <img
          alt="鏡文學"
          src="~/assets/mirrorfiction.png"
        />
      </a>
       
      <a
        href="https://www.readr.tw/"
        target="_blank"
      >
        <img
          alt="READr 讀+"
          src="~/assets/readrlogo-gary.png"
        />
      </a>
    </div>
  </div>
</section>
`;
